Mathlib/Order/WellFounded.lean

```diff
@@ -18,6 +18,35 @@ and provide a few new definitions: `WellFounded.min`, `WellFounded.sup`, and `We
 and an induction principle `WellFounded.induction_bot`.
 -/
 
+theorem acc_def {α} {r : α → α → Prop} {a : α} : Acc r a ↔ ∀ b, r b a → Acc r b where
+  mp h := h.rec fun _ h _ ↦ h
+  mpr := .intro a
+
+theorem exists_not_acc_lt_of_not_acc {α} {a : α} {r} (h : ¬Acc r a) : ∃ b, ¬Acc r b ∧ r b a := by
+  rw [acc_def] at h
+  push_neg at h
+  simpa only [and_comm]
+
+theorem not_acc_iff_exists_descending_chain {α} {r : α → α → Prop} {x : α} :
+    ¬Acc r x ↔ ∃ f : ℕ → α, f 0 = x ∧ ∀ n, r (f (n + 1)) (f n) where
+  mp hx := let f : ℕ → {a : α // ¬Acc r a} :=
+      Nat.rec ⟨x, hx⟩ fun _ a ↦ ⟨_, (exists_not_acc_lt_of_not_acc a.2).choose_spec.1⟩
+    ⟨(f · |>.1), rfl, fun n ↦ (exists_not_acc_lt_of_not_acc (f n).2).choose_spec.2⟩
+  mpr h acc := acc.rec
+    (fun _x _ ih ⟨f, hf⟩ ↦ ih (f 1) (hf.1 ▸ hf.2 0) ⟨(f <| · + 1), rfl, fun _ ↦ hf.2 _⟩) h
+
+theorem acc_iff_isEmpty_descending_chain {α} {r : α → α → Prop} {x : α} :
+    Acc r x ↔ IsEmpty { f : ℕ → α // f 0 = x ∧ ∀ n, r (f (n + 1)) (f n) } := by
+  rw [← not_iff_not, not_isEmpty_iff, nonempty_subtype]
+  exact not_acc_iff_exists_descending_chain
+
+/-- A relation is well-founded iff it doesn't have any infinite descending chain.
+
+See `RelEmbedding.wellFounded_iff_isEmpty` for a version on strict orders. -/
```

Comment:
```suggestion
See `RelEmbedding.wellFounded_iff_isEmpty` for a version in terms of relation embeddings. -/
```
optional, but to me this is the real distinction between the two: the actual spelling of the condition